File: CSharp/CommandMessenger/Queue/ListQueue.cs
using System.Collections.Generic;

namespace CommandMessenger
{
    public class ListQueue<T> : List<T>
    {

        public void Enqueue(T item)
        {
            Add(item);
        }

        public T Dequeue()
        {
            var t = base[0];
            RemoveAt(0);
            return t;
        }

        public T Peek()
        {
            return base[0];
        }

    }
}

File: extras/TypeScript/src/queue/listQueue.ts
export class ListQueue<T> {
  private readonly items: T[] = [];

  get count(): number {
    return this.items.length;
  }

  get isEmpty(): boolean {
    return this.items.length === 0;
  }

  enqueue(item: T): void {
    this.items.push(item);
  }

  enqueueFront(item: T): void {
    this.items.unshift(item);
  }

  dequeue(): T | undefined {
    return this.items.shift();
  }

  peek(): T | undefined {
    return this.items[0];
  }

  clear(): void {
    this.items.length = 0;
  }

  removeAt(index: number): T | undefined {
    if (index < 0 || index >= this.items.length) {
      return undefined;
    }
    return this.items.splice(index, 1)[0];
  }

  findIndex(predicate: (item: T) => boolean): number {
    return this.items.findIndex(predicate);
  }

  toArray(): T[] {
    return [...this.items];
  }
}
